### PR TITLE
Feature/chat

### PR DESCRIPTION
**Pull request by bonham000**
**This PR:**

- Implements Redis cache to store chat history
- Adds Immutable on client to manage chat state
- Improves chat metadata

You will need to `npm install` again and you will also need Redis running locally. [You can download from here](https://redis.io/topics/quickstart) and then run `redis-server` on the command line to start up the Redis server. Should be fairly straightforward (may also be able to install Redis using bower).

Proposing to move toward Immutable for all Redux state management on client. ⭐️ ⭐️ ⭐️ 

_Note: Let's leave all the backend `socket.io` syntax the way it is for now..._

[**and... demo time**](http://recordit.co/VfHgc4v8Db)

_Second Note: Check out all those commits, what a noob! ... I actually had a hard time rebasing I may have screwed up some of the merges so if anything looks amiss that's probably why_ :sob:

### CODE REVIEW

**Comment by no-stack-dub-sack** [comment]
@bonham000 well no conflicts with the base so that's a start... looking forward to pouring over this. We should probably talk again later tonight because some of this seems a little over my head. Loving the ambition here.

**Comment by bonham000** [comment]
@no-stack-dub-sack sounds good! If you're not around tonight but have a chance to merge this please do. 🚀

**Comment by no-stack-dub-sack** [comment]
@bonham000 I will def be able to at least do that. Reviewing now on lunch...

File: server/package.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
     "mongoose": "^4.8.5",
     "passport": "^0.3.2",
     "passport-github2": "^0.1.10",
+    "redis": "^2.7.1",
```
Comment (no-stack-dub-sack) [inline server/package.json]: What's up with the version? Just curious if there's a reason. 
image

File: src/actions/chat.js [hunk 2]
```diff
@@ -1,16 +1,48 @@
 
+import uuidV1 from 'uuid/v1';
```
Comment (no-stack-dub-sack) [inline src/actions/chat.js]: @bonham000  About this... I'm already using `shortid` for flash messages, we could prob shed a dependency and stick with one or the other. Any reason to go with this over `shortid`? 

shortids are shorter I guess so theoretically could repeat sooner, but somehow I doubt that will be an issue.

Comment (no-stack-dub-sack) [inline src/actions/chat.js]: Oh, this is perhaps the diff of the node package and the actual redis itself?

File: server/package.json (hunk 1 above)
Comment (no-stack-dub-sack) [inline server/package.json]: Oh, this is perhaps the diff of the node package and the actual redis itself?

**Pull request merged**